_overviews/scala3-macros/tutorial/reflection.md

```diff
@@ -31,11 +41,126 @@ As it is a subtype of `Tree` we can also look into the `TreeMethods` to find mor
+We can also transform it back into a `Type[?]` using the `asType` method.
+As the type of `Type[?]` is not statically know we need to name it with an existential type to use it, can be done using the `'[t]` pattern.
+
+```scala
+def g[T: Type](using Quotes) =
+  import quotes.reflect.*
+  val tpe: TypeRepr = TypeRepr.of[T]
+  tpe.asType match
+    case '[t] => '{ val x: t = ${...} }
+  ...
+```
+
+## Macro API design
+
+It will be often useful to create helper methods or extractors that to perform some common logic of your macros.
+
+The simples methods will be those that only mention `Expr`, `Type` and `Quotes` in their signature.
+Internally they may use reflection but this will not be seen at the use site of the method.
+
+```scala
+def f(x: Expr[Int])(using Quotes): Expr[Int] =
+  import quotes.reflect.*
+  ...
+```
+
+In some cases it is inevitable to require some methods that work on `Tree`s or other types in `quotes.reflect`.
+For these cases, the best is to follow the following example of method signatures.
 
+A method that takes a `quotes.reflect.Term` parameter
+```scala
+def f(using Quotes)(term: quotes.reflect.Term): String =
+  import quotes.reflect.*
+  ...
+```
+
+An extension method for a `quotes.reflect.Term` returning a `quotes.reflect.Tree`
+```scala
+extension (using Quotes)(term: quotes.reflect.Term)
+  def g: quotes.reflect.Tree = ...
+```
+
+An extractor that matches on `quotes.reflect.Term`s.
+```scala
+object MyExtractor:
+  def unapply(using Quotes)(x: quotes.reflect.Term) =
+    ...
+    Some(y)
+```
+
+> **Avoid saving the `Quotes` context in a field.**
+> `Quotes` in fields inevitably make its use harder by hitting errors involving `Quotes` with different paths.
+>
+> Usually these patterns have been seen in code that uses the Scala 2 ways to define extension methods or contextual unapplies.
+> Now that we have `given` parameters that can be added before other parameters, all these old workarounds are no needed.
+> The new abstraction make it simpler on the definition site, use site and also have a better performance.
+
+## Debugging
+
+### Runtime checks
+
+Unlike `Expr`, it is not possible to guaranty statically that the generated code will be well-formed.
+Hence these checks will be done at runtime (i.e. compile-time when the macro expands).
+
+It is recommended to enable the `-Xcheck-macros` flag while developing macros or on the tests for the macro.
+This flag will enable extra runtime checks that will try to find ill-formed trees or types as soon as they are created.
+
+There is also the `-Ycheck:all` flag that checks all compiler invariants for tree well-formedness.
+These check will usually fail with an assertion error.
+
+### Printing the trees
+
+The `toString` of types in `quotes.reflect` is no great for debugging as they show the internal representation rather than the `quotes.reflect` representation.
```

Comment:
```suggestion
The `toString` methods of types in `quotes.reflect` are not great for debugging as they show the internal representation rather than the `quotes.reflect` representation.
```